src/controllers/user-data-manager.js

```diff
@@ -0,0 +1,132 @@
+    let rs = this._remoteStorageAdapter(data.dataObj.constructor.name)
+    let updatedLocal = await ls.update(data.dataObj,data.params)
+    let updatedRemote = await rs.update(data.dataObj,data.params)
+    // TODO error handling upon update failure
+    return updatedLocal && updatedRemote
+  }
+
+  /**
+   * Delete a single data model object from the user data stores
+   * @param {Object} data object adhering to
+   *                      { dataObj: the data model object to be updated} }
+   * @return {Boolean} true if delete succeeded false if not
+   */
+  async delete(data) {
+    let ls = this._localStorageAdapter(data.dataObj.constructor.name)
+    let rs = this._remoteStorageAdapter(data.dataObj.constructor.name)
+    let deletedLocal = await ls.deleteOne(data.dataObj)
+    let deletedRemote = await rs.deleteOne(data.dataObj)
+    // TODO error handling upon delete failure
+    return deletedLocal && deletedRemote
+  }
+
+  /**
+   * Delete a set objects from the data store
+   * @param {Object} data object adhering to
+   *                      { dataType: the name of the datatype to delete,
+   *                        params: parameters to identify items to be deleted
+   *                      }
+   */
+  async deleteMany(data) {
+    let remoteAdapter =  this._remoteStorageAdapter(data.dataType)
+    let localAdapter = this._localStorageAdapter(data.dataType)
+    localAdapter.deleteMany(data.params)
+    remoteAdapter.deleteMany(data.params)
+  }
+
+  /**
+   * Query the user data stores
+   * @param {Object} data object adhering to
+   *                      { dataType: the name of the datatype to query
+   *                        params: query parameters to
+   *                      }
+   * @return {Object[]} an array of data items
+   */
+  async query(data) {
+    // query queries both the remote and local stores and merges
+    // the results
+    let remoteAdapter =  this._remoteStorageAdapter(data.dataType)
+    let localAdapter = this._localStorageAdapter(data.dataType)
+    let remoteDataItems = await remoteAdapter.query(data.params)
+    let localDataItems = await localAdapter.query(data.params)
+
+    // if we have any remoteData items then we are going to
+    // reset the local store from the remoteData, adding back in any
+    // items that appeared only in the local
+    if (remoteDataItems.length > 0) {
+        localAdapter.deleteMany(params)
+    }
+    let addToRemote = []
+    let updateInRemote = []
+    localDataItems.forEach(item => {
+      let inRemote = false
+      for (let i=0; i<remoteDataItems.length; i++ ) {
+        if (remoteDataItems[i].isSameItem(item)) {
+          inRemote = true
+          // if the item exists in the remote db, check to see if they differ
+          // and if so merge and update
+          if (remoteDataItems[i].isNotEqual(item)) {
+            let merged = remoteDataItems[i].merge(item)
+            remoteDataItems[i] = merged
```

Comment:
I suspect this code will need refinement as we actually introduce the remote store, but for now I went with the simplest approach of having the remote store always just completely overwrite the local store. On lines 89-91, after retrieving the local items I delete them from the local store if there are any in the remote store. My thinking is that the local store may often just be a subset of the remote store, since different devices will have different storage limitations.  So rather than merging locally, I just update the local store with the merged results in the remote store.

Reply:
That makes sense!